### PR TITLE
bugfix: update no longer tries to put non-test watches into the namespace

### LINKED ISSUE #4438: SqliteQueryException in UCM when running 'update' with only watch expressions in scratch.u
The following exception is generated when running `update` after loading a .u file with only watch expressions.

The attached transcript reproduces the issue.

[update-exception-transcript.md](https://github.com/unisonweb/unison/files/13495559/update-exception-transcript.md)

Output is:

```
SqliteQueryException {sql = "SELECT object.id FROM object JOIN hash ON object.primary_hash_id = hash.id WHERE hash.base32 = ? COLLATE NOCASE", params = [SQLText "dfs8orl88dbd401n1jhqttsltaqdtfiuddb16u3vtclhlljg2eslt7q7b1af5lmk0ef8geo2hqn7ng8buadh5ekncllpes044phd19g"], exception = ExpectedExactlyOneRowException {rows = "[]"}, callStack = [], connection = Connection { name = "main", file = "/Users/richardc/.unison/v2/unison.sqlite3" }, threadId = ThreadId 884}
```

Unison installed via brew on MacOS 12.7.1:
```
❯ ucm version
ucm version: release/M5i (built on 2023-11-28)

❯ uname -a
Darwin Christines-MacBook-Pro.local 21.6.0 Darwin Kernel Version 21.6.0: Wed Oct  4 23:55:28 PDT 2023; root:xnu-8020.240.18.704.15~1/RELEASE_X86_64 x86_64

❯ sw_vers
ProductName:	macOS
ProductVersion:	12.7.1
BuildVersion:	21G920
```

### PR DESCRIPTION
**Pull request by mitchellwrosen**
## Overview

Fixes #4438 

## Implementation notes

`update` no longer tries to put all terms found in a Unison file into the namespace: just those that either don't have a watch kind, or those whose watch kind is `"test"`.

## Test coverage

Transcript included.

## Loose ends

The same bug still exists in `add` and `update.old`. Fixing it in `update` unfortunately didn't fix it over there for free.

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase.hs [hunk 1]
```diff
@@ -328,9 +328,7 @@ addDefsToCodebase c uf = do
   traverse_ goTerm (UF.hashTermsId uf)
   where
     goTerm t | debug && trace ("Codebase.addDefsToCodebase.goTerm " ++ show t) False = undefined
-    goTerm (_, r, Nothing, tm, tp) = putTerm c r tm tp
-    goTerm (_, r, Just WK.TestWatch, tm, tp) = putTerm c r tm tp
-    goTerm _ = pure ()
+    goTerm (_, r, wk, tm, tp) = when (WK.watchKindShouldBeStoredInDatabase wk) (putTerm c r tm tp)
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Codebase.hs:331]: refactor, no change in behavior

File: unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs [hunk 2]
```diff
@@ -257,11 +255,14 @@ typecheckedUnisonFileToBranchUpdates abort getConstructors tuf = do
       tuf
         & UF.hashTermsId
         & Map.toList
-        & foldMap \(var, (_, ref, _, _, _)) ->
-          let split = splitVar var
-           in [ BranchUtil.makeAnnihilateTermName split,
-                BranchUtil.makeAddTermName split (Referent.fromTermReferenceId ref) Map.empty
-              ]
+        & foldMap \(var, (_, ref, wk, _, _)) ->
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs:258]: this is the bugfix. before, we didn't look at `wk`. now, we do.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs [hunk 3]
```diff
@@ -425,10 +426,17 @@ incrementLastSegmentChar (ForwardName segments) =
...
 
+-- @getTermAndDeclNames file@ returns the names of the terms and decls defined in a typechecked Unison file.
 getTermAndDeclNames :: (Var v) => TypecheckedUnisonFile v a -> Defns (Set Name) (Set Name)
-getTermAndDeclNames tuf = Defns (terms <> effectCtors <> dataCtors) (effects <> datas)
+getTermAndDeclNames tuf =
+  Defns (terms <> effectCtors <> dataCtors) (effects <> datas)
   where
-    terms = keysToNames $ UF.hashTermsId tuf
+    terms =
+      UF.hashTermsId tuf
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs:435]: there was a similar issue here in just extracting type/term names; I'm not sure if it directly contributed to the crash, but we simply don't want watches that aren't "test" watches here, either.

**Review by aryairani — APPROVED** [review]
Nice!

**Pull request merged**